### PR TITLE
Introduces `StateTransition` to `DPC` struct

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

Introduces `StateTransition` to `DPC` struct. This PR does not require resampling parameters.
- Updates `DPC::authorize` to reason over a state transition object
- Adds `ComputeKey` to the account model
- Updates `DPC::execute` to use compute keys for execution
- Updates `PrivateKey` to return `compute_key()` for execution
- Updates `genesis`, `testnet1_integration`, and `testnet2_integration` with state transition approach

## Test Plan

- Adds `snarkvm_dpc::state::Input` test for `new_noop`
- Adds `snarkvm_dpc::state::Output` test for `new_noop`
- Adds `snarkvm_dpc::state::StateTransition` test for `new_coinbase`
- Adds `snarkvm_dpc::state::StateBuilder` test for `add_noop_input` and `add_noop_output`

### CODE REVIEW

File: dpc/src/account/address.rs [hunk 1]
```diff
@@ -82,8 +82,8 @@ impl<C: Parameters> Address<C> {
     }
 
     /// Returns the address as an encryption public key.
-    pub fn to_encryption_key(&self) -> &<C::AccountEncryptionScheme as EncryptionScheme>::PublicKey {
-        &self.encryption_key
+    pub fn to_encryption_key(self) -> <C::AccountEncryptionScheme as EncryptionScheme>::PublicKey {
```
Comment (Pratyush) [inline dpc/src/account/address.rs:85]: Is the `PublicKey` `Copy`? If not, we should rename this to `into`, as per the API guidelines

File: dpc/src/dpc.rs [hunk 2]
```diff
@@ -88,106 +88,67 @@ impl<C: Parameters> DPCScheme<C> for DPC<C> {
...
+            let private_key = match noop_private_key {
+                Some(noop_private_key) => noop_private_key,
+                None => {
+                    index += 1;
+                    &private_keys[index]
+                }
+            };
+
+            // Construct the prover key.
+            prover_keys.push((
```
Comment (Pratyush) [inline dpc/src/dpc.rs:121]: This can clash with a SNARK prover key. Maybe `transaction_keys`?

Comment (howardwu) [reply dpc/src/dpc.rs:121]: Will change to compute keys

File: dpc/src/account/address.rs (hunk 1 above)
Comment (howardwu) [reply dpc/src/account/address.rs:85]: Good catch

**Review by raychu86 — APPROVED** [review]

File: dpc/src/state/builder.rs [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    ///
+    /// Adds the given input into the builder.
+    ///
+    pub fn add_input(mut self, input: Input<C>) -> Self {
+        // Ensure there are no outputs assigned yet, as the builder computes joint serial numbers.
+        if !self.outputs.is_empty() {
+            self.errors.push("Builder cannot add new inputs after outputs".into());
+        }
+
+        match self.inputs.len() < C::NUM_INPUT_RECORDS {
```
Comment (weikengchen) [inline dpc/src/state/builder.rs:58]: Can the number of inputs "==" to `C::NUM_INPUT_RECORDS`?

File: dpc/src/state/builder.rs [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+            self = self.add_input(input);
+        }
+        self
+    }
+
+    ///
+    /// Adds the given output into the builder.
+    ///
+    pub fn add_output(mut self, output: Output<C>) -> Self {
+        match self.outputs.len() < C::NUM_OUTPUT_RECORDS {
```
Comment (weikengchen) [inline dpc/src/state/builder.rs:79]: Same as above, can the number of outputs "==" `C::NUM_OUTPUT_RECORDS`?

File: dpc/src/state/builder.rs [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+        }
+        self
+    }
+
+    ///
+    /// Appends the given data to the memo field in the builder.
+    ///
+    pub fn append_memo(mut self, data: &Vec<u8>) -> Self {
+        // TODO (howardwu): Change this to not be hardcoded to 64.
+        match self.memo.len() < 64 && (self.memo.len() + data.len()) < 64 {
```
Comment (weikengchen) [inline dpc/src/state/builder.rs:101]: Is is safe to have them "=="?

**Review by weikengchen — APPROVED** [review]
The code looks good. My only question is the one above: regarding the requirement that the number of inputs and outputs to be "<" than the limit, and why our tests do not hit this corner case.

File: dpc/src/state/builder.rs (hunk 3 above)
Comment (howardwu) [reply dpc/src/state/builder.rs:58]: It cannot equal, that would exceed then

File: dpc/src/state/builder.rs (hunk 4 above)
Comment (howardwu) [reply dpc/src/state/builder.rs:79]: Same, it would exceed

File: dpc/src/state/builder.rs (hunk 5 above)
Comment (howardwu) [reply dpc/src/state/builder.rs:101]: Good catch the latter case can, the former cannot

Comment (howardwu) [reply dpc/src/state/builder.rs]: Fixed

File: dpc/src/state/builder.rs (hunk 3 above)
Comment (weikengchen) [reply dpc/src/state/builder.rs:58]: Can you provide more info? Since intuitively it looks like a user can provide all the input records (i.e., no noop padding)?

Comment (weikengchen) [reply dpc/src/state/builder.rs:58]: Is there a compulsory padding or?

Comment (howardwu) [reply dpc/src/state/builder.rs:58]: This `<` just enforces there is at least 1 more input slot available
